pkg/querier/http_test.go

```diff
@@ -34,3 +38,150 @@ func TestTailHandler(t *testing.T) {
 	require.Equal(t, http.StatusBadRequest, rr.Code)
 	require.Equal(t, "multiple org IDs present\n", rr.Body.String())
 }
+
+type slowConnectionSimulator struct {
+	sleepFor   time.Duration
+	deadline   time.Duration
+	didTimeout bool
+}
+
+func (s *slowConnectionSimulator) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	time.Sleep(s.sleepFor)
+	ctx := r.Context()
+
+	select {
+	case <-ctx.Done():
+		switch ctx.Err() {
+		case context.DeadlineExceeded:
+			s.didTimeout = true
+		case context.Canceled:
+			panic("lol")
+		}
+	case <-time.After(s.deadline):
+	}
+}
+
+func TestQueryWrapperMiddleware(t *testing.T) {
+	tenant.WithDefaultResolver(tenant.NewMultiResolver())
+	shortestTimeout := time.Millisecond * 5
+
+	t.Run("request timeout is the shortest one", func(t *testing.T) {
+		defaultLimits := defaultLimitsTestConfig()
+		limits, err := validation.NewOverrides(defaultLimits, nil)
+		require.NoError(t, err)
+		api := NewQuerierAPI(mockQuerierConfig(), nil, limits, log.NewNopLogger())
+
+		// request timeout is 5ms but it sleeps for 100ms, so timeout injected in the request is expected.
+		connSimulator := &slowConnectionSimulator{
+			sleepFor: time.Millisecond * 100,
+			deadline: shortestTimeout,
+		}
+
+		api.cfg.QueryTimeout = time.Millisecond * 10
+		midl := WrapQuerySpanAndTimeout("mycall", api).Wrap(connSimulator)
+
+		req, err := http.NewRequest("GET", "/loki/api/v1/label", nil)
+		ctx, cancelFunc := context.WithTimeout(user.InjectOrgID(req.Context(), "fake"), shortestTimeout)
+		defer cancelFunc()
+		req = req.WithContext(ctx)
+		require.NoError(t, err)
+
+		rr := httptest.NewRecorder()
+		srv := http.HandlerFunc(midl.ServeHTTP)
+
+		srv.ServeHTTP(rr, req)
+		require.Equal(t, http.StatusOK, rr.Code)
+
+		select {
+		case <-ctx.Done():
+			break
+		case <-time.After(shortestTimeout):
+			require.FailNow(t, "should have timed out before %s", shortestTimeout)
+		default:
+			require.FailNow(t, "timeout expected")
+		}
+
+		require.True(t, connSimulator.didTimeout)
+	})
+
+	t.Run("old querier:query_timeout is configured to supersede all others", func(t *testing.T) {
+		defaultLimits := defaultLimitsTestConfig()
+		defaultLimits.QueryTimeout = model.Duration(shortestTimeout)
+		limits, err := validation.NewOverrides(defaultLimits, nil)
+		require.NoError(t, err)
+		api := NewQuerierAPI(mockQuerierConfig(), nil, limits, log.NewNopLogger())
+
+		// configure old querier:query_timeout parameter.
+		// although it is longer than the limits timeout, it should supersede it.
+		api.cfg.QueryTimeout = time.Millisecond * 100
+
+		// querier:query_timeout is 5ms but it sleeps for 100ms, so timeout injected in the request is expected.
```

Comment:
`querier:query_timeout` and `sleepFor` both are `100ms`, no? Am I missing something?

Reply:
sorry, the behavior/test is correct but the comment is wrong. my intention with this test was:
- the new query_timeout is 5ms, but it should be ignored/superseded by the old query_timeout (100ms in this case). This is ensured by `time.After(5ms)` happening before `ctx.Done()`, hence breaking the `select` loop
- Hmm now I'm thinking in adding an additional behavior/mechanism to `connSimulator`: a threshold where it should check that it won't timeout before that threshold. For instance: for this test, the conn simulator will first sleep for 5ms and ensure that it didn't timeout. Then, sleep for 95ms and ensure that it did timeout. WDYT?
-

Reply:
I think you can add just a negative testcase with `dealine` > `sleepFor` and look for `didTimeout` being `false`? It will have the same effect and keep the code simple?

Reply:
good call, thank you. I have though to check for `didTimeout` being truth, right?